pkg/controller/deployment/deployment_controller.go

```diff
@@ -335,8 +357,16 @@ func (dc *DeploymentController) deletePod(obj interface{}) {
 		}
 	}
 	if d := dc.getDeploymentForPod(pod); d != nil && d.Spec.Strategy.Type == extensions.RecreateDeploymentStrategyType {
-		podList, err := dc.listPods(d)
-		if err == nil && len(podList.Items) == 0 {
+		// Sync if this Deployment now has no more Pods.
+		rsList, err := dc.getReplicaSetsForDeployment(d)
```

Comment:
As discussed in the other thread, I'm now in agreement that Deployment should only care about Pods that match its selector. However, I still think it's important in this function to filter both by Deployment selector AND by ControllerRef matching a ReplicaSet that we own. Otherwise, we will count Pods that we shouldn't be counting, and could get stuck waiting.

Is that what you have in mind?